packages/typescript-estree/tools/test-utils.ts

```diff
@@ -84,3 +68,67 @@ export function isJSXFileType(fileType: string): boolean {
   }
   return fileType === 'js' || fileType === 'jsx' || fileType === 'tsx';
 }
+
+/**
+ * Removes the given keys from the given AST object recursively
+ * @param root A JavaScript object to remove keys from
+ * @param keysToOmit Names and predicate functions use to determine what keys to omit from the final object
+ * @param nodes advance ast modifications
+ * @returns formatted object
+ */
+export function omitDeep<T = Record<string, unknown>>(
+  root: T,
+  keysToOmit: { key: string; predicate: (value: unknown) => boolean }[] = [],
+  nodes: Record<string, (node: T, parent: T | null) => void> = {},
+): T {
+  function isObjectLike(value: unknown | null): value is T {
+    return (
+      typeof value === 'object' && !(value instanceof RegExp) && value !== null
+    );
+  }
+
+  function shouldOmit(keyName: string, val: unknown): boolean {
+    if (keysToOmit?.length) {
+      return keysToOmit.some(
+        keyConfig => keyConfig.key === keyName && keyConfig.predicate(val),
+      );
+    }
+    return false;
+  }
+
+  function visit(oNode: T, parent: T | null): T {
```

Comment:
This shouldn't be tied to the parent generic, because it makes it look like there are no mutations applied.
Doing so is very, very unsafe for obvious reasons.

You should type this as the generic object-like type to be clear there is no such guarantee.

```suggestion
  function visit(oNode: Record<string, unknown>, parent: Record<string, unknown> | null): Record<string, unknown> {
```